### ISSUE TITLE
Configure auth model and configure for use with Typescript

### ISSUE THREAD

**Issue by rodriggj** [rodriggj/ticketing_app#33]

**Comment by rodriggj** [comment]
### Intended signup flow 
<p align="center">
<img width="300" height="275" alt="image" src="https://user-images.githubusercontent.com/8760590/162620268-1bf4c435-b80b-4fd8-86b7-160db703d798.png">
</p>

**Comment by rodriggj** [comment]
Type script will want to validate the Object definition for our User creation which will cause 2 issues we will have to address in our Model and Schema definition: 

1. Typescript will want to have a definition of each field we pass to Mongo 
2. MongoDB will potentially add fields to our Object that Typescript will want to recognize

These will be addressed, just be aware they need to be managed.

**Comment by rodriggj** [comment]
1. Start by creating a `models` dir under `auth / src` subdirectory
2. Create a file in this dir called `Users.ts`
3. In this file input the following code: 
```js
import mongoose from 'mongoose'

const userSchema = new mongoose.Schema({
    email: {
        type: String, 
        required: true
    }, 
    password: {
        type: String, 
        required: true
    }
})

const User = mongoose.model('User', userSchema)

export { User }
```

**Comment by rodriggj** [comment]
### Fix to Issue # 1

```js
import mongoose from 'mongoose'

// Fix to issue1
interface UserAttrs {
    email: string, 
    password: string
}

const userSchema = new mongoose.Schema({
    email: {
        type: String, 
        required: true
    }, 
    password: {
        type: String, 
        required: true
    }
})

const User = mongoose.model('User', userSchema)

// Fix to issue1
const buildUser = (attrs: UserAttrs) => {
    return new User(attrs)
}

export { User, buildUser }
```

**Comment by rodriggj** [comment]
### Fix to Issue #2 

```js
import mongoose from 'mongoose'

// Fix to issue1
// An interface that describes the properties that are required to create a new User
interface UserAttrs {
    email: string, 
    password: string
}

// Refactor to issue1 
// An interface that describes the properties that a new User Model has
interface UserModel extends mongoose.Model<UserDoc>{
    build(attrs: UserAttrs): UserDoc
}

// Fix to issue2
// An interface that describes the properties that a User Document has
interface UserDoc extends mongoose.Document{
    email: string, 
    password: string
}

const userSchema = new mongoose.Schema({
    email: {
        type: String, 
        required: true
    }, 
    password: {
        type: String, 
        required: true
    }
})
userSchema.statics.build = (attrs: UserAttrs) => {
    return new User(attrs)
}

const User = mongoose.model<UserDoc, UserModel>('User', userSchema)

export { User } 
```

**Comment by rodriggj** [comment]
Now that the Model is built and can handle TypeScript `Generics` ... lets wire up the User Model to the SignUp route

1. Nav to the `routes/signup.ts` file
2. Import the `User Model`
3. Update the file with the following code: 
```js
import express, { Request, Response }from 'express'
import { body, validationResult } from 'express-validator'
import { User } from '../models/users'
import { RequestValidationError } from '../errors/request-validation-error'

const router = express.Router()

router.post(
    '/api/users/signup', 
[
    body('email')
      .isEmail()
      .withMessage('Email must be valid'),
    body('password')
      .trim()
      .isLength({min:4, max:20})
      .withMessage('Password must be between 4 and 20 characters')
], 
async (req: Request, res: Response) => {
        const errors = validationResult(req)

        if(!errors.isEmpty()) {
            throw new RequestValidationError(errors.array())
        }

        const { email, password } = req.body

        // Query existing User Documents to see if User already exists
        const existingUser = await User.findOne({ email })
        if(existingUser) {
            console.log('Email already exists.')
            return res.send({})
        }

        // If User does not exist we will Hash Password

        // We will then create the User Document in mongodb
        const user = User.build({ email, password })
        await user.save()

        // Send back to the user some token for Auth

        res.status(201).send(user)
    }
)

export { router as signupRouter}
```

> NOTE: Password Hash and Send Token steps in our process will be completed in subsequent steps.
